spec/polymorphic_spec.cr

```diff
@@ -0,0 +1,123 @@
+
+private class OptionalPolymorphicEvent < BaseModel
+  table :polymorphic_events do
+    belongs_to task : PolymorphicTask?
+    belongs_to task_list : PolymorphicTaskList?
+    polymorphic :optional_eventable, optional: true, associations: [:task, :task_list]
+  end
+end
+
+describe "polymorphic belongs to" do
+  it "sets up a method for accessing associated record" do
+    task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+    event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+    event.eventable.should eq(task)
+
+    task_list = PolymorphicTaskList::SaveOperation.create!(title: "Use Lucky")
+    event = PolymorphicEvent::SaveOperation.create!(task_list_id: task_list.id)
+    event.eventable.should eq(task_list)
+  end
+
+  it "can require preloading" do
+    with_lazy_load(enabled: false) do
+      expect_raises Avram::LazyLoadError do
+        task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+        event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+        event.eventable # should raise because it was not preloaded
+      end
+    end
+  end
+
+  it "has ! method to allow lazy loading" do
+    with_lazy_load(enabled: false) do
+      task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+      event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+      event.eventable!.should eq(task)
+    end
+  end
+
+  it "can preload the polymorphic associations" do
+    with_lazy_load(enabled: false) do
+      task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+      event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+      event = PolymorphicEvent::BaseQuery.new.preload_eventable.find(event.id)
+      event.eventable.should eq(task)
+
+      # Check that it preloads both belongs_to, not just the first
+      task_list = PolymorphicTaskList::SaveOperation.create!(title: "Use Lucky")
+      event = PolymorphicEvent::SaveOperation.create!(task_list_id: task_list.id)
+      event = PolymorphicEvent::BaseQuery.new.preload_eventable.find(event.id)
+      event.eventable.should eq(task_list)
+    end
+  end
+
+  describe "required (default)" do
+    it "validates that exactly one polymorphic belongs to is allowed" do
+      operation = PolymorphicEvent::SaveOperation.create(task_id: 1, task_list_id: 1) do |operation, _event|
+        operation.valid?.should be_false
+        operation.task_list_id.errors.should eq(["must be blank"])
+      end
+
+      operation = PolymorphicEvent::SaveOperation.create do |operation, _event|
+        operation.valid?.should be_false
+        operation.task_id.errors.should eq(["at least one 'eventable' must be filled"])
+      end
+    end
+
+    it "type should not include nil" do
+      task = PolymorphicTask::SaveOperation.create!(title: "Use Lucky")
+      event = PolymorphicEvent::SaveOperation.create!(task_id: task.id)
+      typeof(event.eventable).should eq(PolymorphicTask | PolymorphicTaskList)
```

Comment:
You could also use `be_a` (I think)

```crystal
event.eventable.should be_a(PolymorphicTask | PolymorphicTaskList)
```

Reply:
I will give it a try!

Reply:
Ok so `be_a` is for a single type as it operates on the runtime type and not compile time type. An example probably works best to describe this:

```crystal
def mixed_array
  [1, "foo"]
end

typeof(mixed_array) # Will be Array(String | Int32). typeof works at compile time. Does not actually call the method.
mixed_array.first # Will be String | Int32 at compile time, but at runtime it is one of the two, not *both*
mixed_array.first.should be_a(Int32) # We can only pass one type
```

Reply:
Interesting. Thanks for trying!